schemas/Clinical/index.js

```diff
@@ -0,0 +1,204 @@
+    specimenId: String!
+  }
+
+  type Query {
+    """
+    Retrieve current stored Clinical Registration data for a program
+    """
+    clinicalRegistration(shortName: String!, file: Upload!): ClinicalRegistrationData!
+  }
+
+  type Mutation {
+    """
+    Upload a Registration file
+    """
+    uploadClinicalRegistration(
+      shortName: String!
+      registrationFile: Upload!
+    ): ClinicalRegistrationData!
+
+    """
+    Remove the Clinical Registration data currently uploaded and not committed
+    """
+    clearClinicalRegistration(shortName: String!, registrationId: String!): Boolean!
+
+    """
+    Complete registration of the currently uploaded Clinical Registration data
+    """
+    commitClinicalRegistration(shortName: String!, registrationId: String!): Boolean!
+  }
+`;
+
+const convertRegistrationRecordToGql = (record, row) => ({
+  row,
+  programShortName: record.program_id,
+  donorSubmitterId: record.submitter_donor_id,
+  gender: record.gender,
+  specimenSubmitterId: record.submitter_specimen_id,
+  specimentType: record.specimen_type,
+  tumourNormalDesignation: record.tumour_normal_designation,
+  sampleSubmitterId: record.submitter_sample_id,
+  sampleType: record.sample_type,
+});
+
+const convertRegistrationStatsToGql = statsEntry => {
+  const output = {
+    count: 0,
+    rows: [],
+    names: [],
+    values: [],
+  };
+  const names = Object.keys(statsEntry) || [];
+  output.count = names.length;
+  names.forEach(name => {
+    output.names.push(name);
+    const rows = statsEntry[name];
+    rows.forEach(row => !output.rows.includes(row) && output.rows.push(row));
+    output.values.push({ name, rows });
+  });
+
+  return output;
+};
+
+const convertRegistrationErrorToGql = errorData => ({
+  type: errorData.type,
+  message: get(ERROR_MESSAGES, errorData.type, ''), // Empty String if we don't have a message for the given type
+  row: errorData.index,
+  field: errorData.fieldName,
+  value: errorData.info.value,
+  sampleId: errorData.info.sampleSubmitterId,
+  donorId: errorData.info.donorSubmitterId,
+  specimenId: errorData.info.specimenSubmitterId,
+});
+
+const convertRegistrationDataToGql = data => {
+  const registration = get(data, 'registration', {});
+  return {
+    id: registration._id || null,
+    creator: registration.creator || null,
+    records: get(registration, 'records', []).map((record, i) =>
+      convertRegistrationRecordToGql(record, i),
+    ),
+    errors: get(data, 'errors', []).map((errorData, i) =>
+      convertRegistrationErrorToGql(errorData, i),
+    ),
+
+    newDonors: convertRegistrationStatsToGql(get(registration, 'stats.newDonorIds', {})),
+    newSpecimens: convertRegistrationStatsToGql(get(registration, 'stats.newSpecimenIds', {})),
+    newSamples: convertRegistrationStatsToGql(get(registration, 'stats.newSampleIds', {})),
+    alreadyRegistered: convertRegistrationStatsToGql(
+      get(registration, 'stats.alreadyRegistered', {}),
+    ),
```

Comment:
seems to be a lot of wrangling going on. I think it's possible make graphql lazy evaluate them by thunking like so (basically make them in-line resolvers): 
```
records: () => get(registration, 'records', []).map((record, i) =>
  convertRegistrationRecordToGql(record, i),
)
```
Give it a shot, if it works I'd recommend that to reduce server load :)

Reply:
Let me look into lazy resolving the fields. I'll update the PR...

Reply:
theoretically, that thunk is just a resolver so it receives the same arguments as the others as well, in case that's ever useful :)